enterprise/backend/src/metabase_enterprise/scim/v2/api.clj

```diff
@@ -1,14 +1,419 @@
+
+(mu/defn ^:private mb-group->scim :- SCIMGroup
+  "Given a Metabase permissions group, returns a SCIM group."
+  [group]
+  {:schemas     [group-schema-uri]
+   :id          (:entity_id group)
+   :members     (map
+                 (fn [member]
+                   {:value   (:entity_id member)
+                    :$ref    (str (scim/scim-base-url) "/Users/" (:entity_id member))
+                    :display (:email member)})
+                 (:members group))
+   :displayName (:name group)
+   :meta        {:resourceType "Group"}})
+
+(defn- group-filter-clause
+  [filter-parameter]
+  (let [[_ match] (re-matches #"^displayName eq \"(.*)\"$" filter-parameter)]
+    (if match
+      [:= :name match]
+      (throw (ex-info "Unsupported filter parameter" {:filter      filter-parameter
+                                                      :status-code 400})))))
+
+(defendpoint GET "/Groups"
+  "Fetch a list of groups."
+  [:as {{start-index :startIndex c :count filter-param :filter} :params}]
+  {start-index  [:maybe ms/PositiveInt]
+   c            [:maybe ms/PositiveInt]
+   filter-param [:maybe ms/NonBlankString]}
+  (let [limit          (or c default-pagination-limit)
+        ;; SCIM start-index is 1-indexed, so we need to decrement it here
+        offset         (if start-index (dec start-index) default-pagination-offset)
+        filter-param   (when filter-param (codec/url-decode filter-param))
+        filter-clause  (if filter-param
+                         (group-filter-clause filter-param)
+                         [:= 1 1])
+        groups         (t2/select (cons :model/PermissionsGroup group-cols)
+                                  {:where    [:and
+                                               [:not= :id (:id perms-group/all-users)]
+                                               [:not= :id (:id perms-group/admin)]
+                                               filter-clause]
+                                   :limit    limit
+                                   :offset   offset
+                                   :order-by [[:id :asc]]})
+        results-count  (count groups)
+        items-per-page (if (< results-count limit) results-count limit)
+        result         {:schemas      [list-schema-uri]
+                        :totalResults (t2/count :model/PermissionsGroup {:where filter-clause})
+                        :startIndex   (inc offset)
+                        :itemsPerPage items-per-page
+                        :Resources    (map mb-group->scim groups)}]
+    (scim-response result)))
+
+(defendpoint GET "/Groups/:id"
+  "Fetch a single group."
+  [id]
+  {id ms/NonBlankString}
+  (-> (get-group-by-entity-id id)
+      (t2/hydrate :scim_group_members)
+      mb-group->scim))
+
+(defn- update-group-membership
+  "Updates the membership of `group-id` to be the set of users in the collection `user-entity-ids`. Clears
+  any existing members. May not be run on the Administrators or All Users groups."
```

Comment:
Is something enforcing the prohibition on running this on the Administrators/All Users groups?

Reply:
Good catch! Fixing

Reply:
I've instead updated `get-group-by-entity-id` to not return the Admin/All Users group so that they're invisible to SCIM entirely